### PR TITLE
Link Preview

### PR DESCRIPTION
**Pull request by MatkoTom**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant
motivation and context. List any dependencies that are required for this change.

Implemented Link Preview functionality. Previews will show when you type a valid URL in the text box. Also, we are now displaying previews in message boxes.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can
reproduce. Please also list any relevant details for your test configuration. All testing was done
manually by following, plainly, functionality requirements. If there are some special edge cases or
special ways that things need to be tested they will be mentioned below.

Dev testing

**Test Configuration**:

* Firmware version:
* Hardware: SAMSUNG Galaxy S22
* SDK: Android 14

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

JPEG_20240314_133227_114252226225663100
JPEG_20240314_133227_5498126908096346632
JPEG_20240314_133227_5675096233419327355
JPEG_20240314_133227_8189280755558129014
JPEG_20240314_133228_6164645997527896083

### CODE REVIEW

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/MainActivity.kt [hunk 1]
```diff
@@ -172,6 +177,27 @@ class MainActivity : BaseActivity() {
...
+            val inputStream = connection.inputStream
+            val reader = BufferedReader(InputStreamReader(inputStream))
+            val stringBuilder = StringBuilder()
+            var line: String?
+            while (reader.readLine().also { line = it } != null) {
+                stringBuilder.append(line).append("\n")
+            }
+            reader.close()
+            connection.disconnect()
+            Timber.d("Page data = $stringBuilder")
```
Comment (ivolarevic) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/MainActivity.kt:199]: If we don't need this log, delete it.

File: app/src/main/res/layout/preview_action.xml [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            app:layout_constraintStart_toEndOf="@id/iv_preview_image"
+            app:layout_constraintTop_toTopOf="parent" />
+
+        <TextView
+            android:id="@+id/tv_description"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginStart="@dimen/sixteen_dp_margin"
+            android:ellipsize="end"
+            android:lineHeight="10dp"
```
Comment (ivolarevic) [inline app/src/main/res/layout/preview_action.xml:56]: Same as above

File: app/src/main/res/layout/preview_layout.xml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+<com.google.android.material.card.MaterialCardView xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:id="@+id/cl_preview"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:backgroundTint="?attr/additionalColor"
+    android:clickable="true"
+    android:focusable="true"
+    android:visibility="visible"
```
Comment (ivolarevic) [inline app/src/main/res/layout/preview_layout.xml:11]: Do we need here visibility?

**Review by ivolarevic — CHANGES_REQUESTED** [review]
Please fix comments and re-submit pr.

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/MainActivity.kt (hunk 1 above)
Comment (MatkoTom) [reply app/src/main/java/com/clover/studio/spikamessenger/ui/main/MainActivity.kt:199]: Method was removed because it has no purpose

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/MainActivity.kt [hunk 4]
```diff
@@ -172,6 +177,27 @@ class MainActivity : BaseActivity() {
         }
 
         viewModel.getUnreadCount()
+        getPageContent()
+    }
+
+    private fun getPageContent() {
+        CoroutineScope(Dispatchers.IO).launch {
+            val connection = withContext(Dispatchers.IO) {
+                URL("https://www.youtube.hr").openConnection()
```
[the remark this answers is not in the document]
Comment (MatkoTom) [reply app/src/main/java/com/clover/studio/spikamessenger/ui/main/MainActivity.kt:186]: Method was removed because it has no purpose

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatAdapter.kt [hunk 5]
```diff
@@ -236,6 +237,22 @@ class ChatAdapter(
                         )
                     }
 
+                    /** Show message preview: */
+                    if (it.message.body?.thumbnailData != null && it.message.body.thumbnailData?.title?.isNotEmpty() == true && it.message.deleted == false) {
+//                        setViewsVisibility(holder.binding.flPreviewMsgContainer, holder)
```
[the remark this answers is not in the document]
Comment (MatkoTom) [reply app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatAdapter.kt:242]: It will be removed

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatAdapter.kt [hunk 6]
```diff
@@ -375,6 +392,22 @@ class ChatAdapter(
...
```
[the remark this answers is not in the document]
Comment (MatkoTom) [reply app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatAdapter.kt:397]: It will be removed

File: app/src/main/res/layout/preview_action.xml [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            android:layout_marginStart="@dimen/sixteen_dp_margin"
+            android:layout_marginEnd="@dimen/sixteen_dp_margin"
+            android:paddingTop="@dimen/four_dp_margin"
+            android:paddingBottom="@dimen/four_dp_margin"
+            android:ellipsize="end"
+            android:fontFamily="@font/roboto_bold"
+            android:maxLines="2"
+            android:text="@string/placeholder_title"
+            android:textAppearance="@style/CustomTextBody"
+            android:lineHeight="10dp"
```
[the remark this answers is not in the document]
Comment (MatkoTom) [reply app/src/main/res/layout/preview_action.xml:39]: It is overriding the style lineHeight which is 20dp. We need 10dp here.

File: app/src/main/java/com/clover/studio/spikamessenger/utils/helpers/ChatAdapterHelper.kt [hunk 8]
```diff
@@ -38,13 +39,18 @@ object ChatAdapterHelper {
             holder.itemView.findViewById<TextView>(R.id.tv_message),
             holder.itemView.findViewById<CardView>(R.id.cv_media),
             holder.itemView.findViewById<FrameLayout>(R.id.fl_reply_msg_container),
+            holder.itemView.findViewById<FrameLayout>(R.id.fl_preview_msg_container)
         )
 
         viewsToHide.forEach {
-            it.visibility = if (it == viewToShow) View.VISIBLE else View.GONE
+            it.visibility = if (it == viewToShow || it.isVisible) View.VISIBLE else View.GONE
```
[the remark this answers is not in the document]
Comment (MatkoTom) [reply app/src/main/java/com/clover/studio/spikamessenger/utils/helpers/ChatAdapterHelper.kt:46]: This needs to be removed

File: app/src/main/res/layout/preview_action.xml [hunk 9]
```diff
@@ -0,0 +1,100 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:id="@+id/cl_message_preview"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:background="?attr/secondAdditionalColor"
+    android:paddingBottom="@dimen/ten_dp_margin"
```
[the remark this answers is not in the document]
Comment (MatkoTom) [reply app/src/main/res/layout/preview_action.xml:9]: It will kill (work)

File: app/src/main/res/layout/preview_action.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (ivolarevic) [reply app/src/main/res/layout/preview_action.xml:39]: Ok, you can extract this

**Review by ivolarevic — APPROVED** [review]
LGTM
